typescript/core/src/parse.ts

```diff
@@ -1,374 +1,415 @@
 import { crc32 } from "@foxglove/crc";
 
 import Reader from "./Reader";
-import { isKnownOpcode, MCAP_MAGIC, Opcode } from "./constants";
+import { MCAP_MAGIC, Opcode } from "./constants";
 import { McapMagic, TypedMcapRecord } from "./types";
 
 /**
  * Parse a MCAP magic string at `startOffset` in `view`.
  */
-export function parseMagic(
-  view: DataView,
-  startOffset: number,
-): { magic: McapMagic; usedBytes: number } | { magic?: undefined; usedBytes: 0 } {
-  if (startOffset + MCAP_MAGIC.length > view.byteLength) {
-    return { usedBytes: 0 };
+export function parseMagic(reader: Reader): McapMagic | undefined {
+  if (reader.bytesRemaining() < MCAP_MAGIC.length) {
+    return undefined;
   }
-  if (!MCAP_MAGIC.every((val, i) => val === view.getUint8(startOffset + i))) {
+  const magic = reader.u8ArrayBorrow(MCAP_MAGIC.length);
+  if (!MCAP_MAGIC.every((val, i) => val === magic[i])) {
     throw new Error(
       `Expected MCAP magic '${MCAP_MAGIC.map((val) => val.toString(16).padStart(2, "0")).join(
         " ",
-      )}', found '${Array.from(MCAP_MAGIC, (_, i) =>
-        view
-          .getUint8(startOffset + i)
-          .toString(16)
-          .padStart(2, "0"),
-      ).join(" ")}'`,
+      )}', found '${Array.from(magic, (_, i) => magic[i]!.toString(16).padStart(2, "0")).join(
+        " ",
+      )}'`,
     );
   }
-  return {
-    magic: { specVersion: "0" },
-    usedBytes: MCAP_MAGIC.length,
-  };
+  return { specVersion: "0" };
 }
 
 /**
- * Parse a MCAP record beginning at `startOffset` in `view`.
+ * Parse a MCAP record from the given reader
  */
-export function parseRecord({
-  view,
-  startOffset,
-  validateCrcs,
-}: {
-  view: DataView;
-  startOffset: number;
-  validateCrcs: boolean;
-}): { record: TypedMcapRecord; usedBytes: number } | { record?: undefined; usedBytes: 0 } {
-  if (startOffset + /*opcode*/ 1 + /*record content length*/ 8 >= view.byteLength) {
-    return { usedBytes: 0 };
+// eslint-disable-next-line @foxglove/no-boolean-parameters
+export function parseRecord(reader: Reader, validateCrcs = false): TypedMcapRecord | undefined {
```

Comment:
there's been a consensus (or ruling?) among the team that naked boolean parameters harm readability at call sites, so if this _needs_ to be this way for performance reasons, we'll need at least a comment to that effect. (I'm not sure how big of an effect the removal of the params object had?)